extensions/amp-video/0.1/mixins.js

```diff
@@ -0,0 +1,524 @@
+
+    listen(element, MediaPoolEvents.DEALLOCATED, e => {
+      const {detail} = e;
+      this.isAllocated_ = false;
+      this.mediaInfo_ =
+        /** @type {!../../amp-story/0.1/media-pool.MediaInfoDef} */ (detail);
+      this.showPlaceholder();
+    });
+
+    listen(element, 'volumechange', () => {
+      setMuted(this.getMediaInfo_('muted'));
+    });
+  }
+
+  /** @visibleForTesting */
+  showPlaceholder() {
+    if (this.placeholder_) {
+      this.togglePlaceholder_(/* show */ true);
+      return this.placeholder_;
+    }
+
+    const {element} = this.impl;
+    const posterSrc = dev().assertString(element.getAttribute('poster'));
+
+    const img = new Image();
+    const {impl} = this;
+    const {win} = this.impl;
+
+    const placeholder = createElementWithAttributes(win.document, 'div', dict({
+      'class': 'i-amphtml-placeholder',
+    }));
+
+    img.src = posterSrc;
+
+    this.placeholder_ = impl.loadPromise(img).then(() =>
+      impl.getVsync().mutatePromise(() => {
+        setImportantStyles(placeholder, {
+          'background-image': `url(${posterSrc})`,
+          'background-size': 'cover',
+          'pointer-events': 'none',
+        });
+        impl.applyFillContent(placeholder);
+        element.appendChild(placeholder);
+        return placeholder;
+      }));
+
+    return this.placeholder_;
+  }
+
+  /** @visibleForTesting */
+  hidePlaceholder() {
+    this.togglePlaceholder_(/* show */ false);
+  }
+
+  /**
+   * @param {boolean} show
+   * @private
+   */
+  togglePlaceholder_(show) {
+    if (!this.placeholder_) {
+      return;
+    }
+    const {impl} = this;
+
+    this.placeholder_.then(placeholder =>
+      impl.getVsync().mutate(() => {
+        if (show) {
+          resetStyles(placeholder, ['display']);
+          return;
+        }
+        this.placeholder_.then(placeholder =>
+          impl.getVsync().mutate(() => {
+            setImportantStyles(placeholder, {'display': 'none'});
+          }));
+      }));
+  }
+
+  /** @override */
+  whenLoadStarts() {
+    if (!this.isAllocated_) {
+      // Return placeholder promise if not allocated so that there is at least
+      // one signal available for `layoutCallback`.
+      return this.showPlaceholder();
+    }
+    const {impl, video} = this;
+    return impl.loadPromise(video);
+  }
+
+  /** @override */
+  play() {
+    const video =
+        /** @type {!HTMLMediaElement} */ (dev().assertElement(this.video));
+    return this.poolPromise_.then(mediaPool =>
+      mediaPool.play(video));
+  }
+
+  /** @override */
+  pause() {
+    const video =
+        /** @type {!HTMLMediaElement} */ (dev().assertElement(this.video));
+    return this.poolPromise_.then(mediaPool =>
+      mediaPool.pause(video));
+  }
+
+  /** @override */
+  toggleMuted(unusedMuted) {
+    // TODO(alanorozco): Implement method in media pool.
+  }
+
+  /** @override */
+  toggleControls(unusedShow) {
+    // TODO(alanorozco): Implement method in media pool.
```

Comment:
For `toggleControls`, `fullscreenEnter`, and `fullscreenExit`, should you `dev().error('MEDIA-POOL', 'unimplemented')`?